File: py/66.py
# 661
import math
N = 1000

m = 0
md = 0
for d in range(N+1):
    x = math.sqrt(d)
    a0 = int(x)
    if x == a0:
        continue
    P = a0
    Q = d - a0 * a0
    a = (a0 + P) / Q
    pp = a0
    pn = a0 * a + 1
    qp = 1
    qn = a
    while pn * pn  - d * qn * qn != 1:
        P = a * Q - P
        Q = (d - P * P) / Q
        a = (a0 + P) / Q
        pp, pn = pn, a * pn + pp
        qp, qn = qn, a * qn + qp
    if pn > m:
        m = pn
        md = d
print md

File: hs/66.hs
-- 661
import Data.Function(on)
import Data.List(maximumBy)
import Euler(expConvergents, sqrtExpansion)

nn = 1000

-- when solving x^2-Dy^2 = 1 (Pell's Equation)
-- wikipedia says the smallest x is found by evaluating
-- the convergents of sqrt(D) until a solution is found

solveDio d
    | null xs = 0
    | otherwise  = fst $ head $ filter isSolution $ expConvergents (n,xs)
    where (n,xs) = sqrtExpansion d
          isSolution (x,y) = x^2 - d*y^2 == 1

maxDio n = maximumBy (compare `on` solveDio) [2..n]

main = putStrLn $ show $ maxDio nn
